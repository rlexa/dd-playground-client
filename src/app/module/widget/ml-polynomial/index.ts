export * from './ml-polynomial.component';
export * from './ml-polynomial.module';
