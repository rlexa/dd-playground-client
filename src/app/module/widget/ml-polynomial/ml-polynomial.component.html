<div ddFlexbox="vys">
  <mat-card ddFlexitem="at">
    <mat-card-title>Polynomial</mat-card-title>
    <mat-card-content>
      <div ddFlexbox="vys">

        <div ddFlexbox="hys" class="childv">

          <table *ngIf="factorsCurrent$ | async as factorsCurrent">
            <tr>
              <td></td>
              <td *ngFor="let _ of factorsCurrent; let iFactor = index; trackBy: trackByIndex">
                <mat-form-field class="width-short-input font-table-input mat-input-save-subspace childh">
                  <input matInput type="number" [min]="-factorRange" [max]="factorRange" [disabled]="isBusy$ | async" required [value]="factorsCurrent[iFactor]"
                    (change)="updateFactor(factorsCurrent, iFactor, +$event.target.value, $event.target)">
                </mat-form-field>
              </td>
            </tr>
            <tr *ngFor="let factors of [factorsCurrent, factorsTrained$ | async]; let iFactors = index;" [ngStyle]="{'color': colors[iFactors % colors.length], 'vertical-align': 'bottom'}">
              <td class="text-nowrap">f(x) = </td>
              <ng-container *ngFor="let factor of factors; let ii = index;">
                <td>
                  {{ (factor >= 0 ? ' +' : ' ') + (factor | d3number:'.2r') }}
                  <ng-container *ngIf="ii+1 < factors.length">
                    {{ 'x' }}
                    <sup *ngIf="ii+2 < factors.length" class="font-small">{{ factors.length - 1 - ii }}</sup>
                  </ng-container>
                </td>
              </ng-container>
            </tr>
          </table>

          <div ddFlexbox="vys" class="childhv">
            <div ddFlexbox="hyscsw" class="childhv">
              <div class="childh">
                <button mat-raised-button (click)="generatePoints()" [disabled]="isBusy$ | async">NEW POINTS</button>
              </div>
              <div class="childh">
                <button mat-raised-button (click)="resetTrained()" [disabled]="isBusy$ | async">RESET TRAINED</button>
              </div>
            </div>
            <div ddFlexbox="hyscsw" class="childhv">
              <div ddFlexbox="hys" class="margin-top">
                <label class="childhv">TFJS:</label>
                <app-simple-view [isDense]="true" [isExpanded]="false" [isExpandable]="true" [data]="{'backend': tfjsBackend$ | async, 'memory': tfjsMemory$ | async}"></app-simple-view>
              </div>
              <mat-form-field class="childh width-short-input font-table-input">
                <input matInput type="number" placeholder="Rate" [disabled]="isBusy$ | async" step=".1" min="0.01" [value]="learningRate$ | async"
                  (change)="updateLearningRate(+$event.target.value)">
              </mat-form-field>
              <mat-form-field class="childh width-shorter-input font-table-input">
                <mat-select placeholder="Optimizer" [value]="(optimizer$ | async) || optimizerDef" (selectionChange)="updateOptimizer($event.value)">
                  <mat-option *ngFor="let ii of optimizers" [value]="ii"> {{ ii }} </mat-option>
                </mat-select>
              </mat-form-field>
              <button *ngFor="let ii of [1, 5, 10, 100, 1000]; let iindex= index;" mat-raised-button class="auto-min-width" (click)="train(ii)"
                [disabled]="isBusy$ | async">
                {{ (iindex === 0 ? 'TRAIN ' : '') + 'x' + ii }}
              </button>
            </div>
          </div>

        </div>
        <app-diagram-polynom [style.minWidth]="'60rem'" ddFlexitem="at" [points]="pointsCurrent$ | async" [polynomials]="[factorsCurrent$ | async, factorsTrained$ | async]"
          [polynomialColors]="colors"></app-diagram-polynom>
      </div>
    </mat-card-content>
  </mat-card>
</div>