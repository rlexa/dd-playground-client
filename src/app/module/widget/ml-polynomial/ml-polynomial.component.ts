import { ChangeDetectionStrategy, Component, OnDestroy, OnInit } from '@angular/core';
import { DEF_OPTIMIZER, detectPolynom, generatePolynomialPoints, OPTIMIZERS } from 'app/ai';
import { ReduxService, ReduxSetUiAiService } from 'app/redux';
import { trackByIndex } from 'app/util';
import { DoneSubject, RxCleanup } from 'dd-rxjs';
import { BehaviorSubject, Subject } from 'rxjs';
import { debounceTime, tap } from 'rxjs/operators';

@Component({
  selector: 'app-ml-polynomial',
  templateUrl: './ml-polynomial.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MlPolynomialComponent implements OnDestroy, OnInit {
  constructor(
    private readonly redux: ReduxService,
    private readonly reduxSet: ReduxSetUiAiService,
  ) { }

  private readonly numPoints = 20;
  private readonly xRange = 1;
  @RxCleanup() private readonly triggerDoPoints$ = new Subject();
  @RxCleanup() private readonly done$ = new DoneSubject();

  readonly factorRange = 10;
  readonly colors = ['rgba(0, 0, 255, 200)', 'rgba(255, 100, 100, 255)'];
  readonly trackByIndex = trackByIndex;
  readonly optimizers = OPTIMIZERS;
  readonly optimizerDef = DEF_OPTIMIZER;
  @RxCleanup() readonly isBusy$ = new BehaviorSubject(false);

  factorsCurrent$ = this.redux.watch(state => state.ui.ai.mlPolynomial.factorsCurrent, this.done$).pipe(tap(() => this.triggerDoPoints$.next()));
  factorsTrained$ = this.redux.watch(state => state.ui.ai.mlPolynomial.factorsTrained, this.done$);
  learningRate$ = this.redux.watch(state => state.ui.ai.mlPolynomial.learningRate, this.done$);
  pointsCurrent$ = this.redux.watch(state => state.ui.ai.mlPolynomial.pointsCurrent, this.done$);
  optimizer$ = this.redux.watch(state => state.ui.ai.mlPolynomial.optimizer, this.done$);

  ngOnDestroy() { }

  ngOnInit() {
    this.triggerDoPoints$.pipe(debounceTime(1)).subscribe(() => this.doGeneratePoints());
  }

  updateFactor = (factors: number[], index: number, val: number, elem: HTMLInputElement) =>
    this.reduxSet.setMlPolynomialFactorsCurrent(
      factors.map((ii, _index) => {
        let ret = ii;
        if (_index === index) {
          ret = Math.min(Math.max(val || 0, -this.factorRange), this.factorRange);
          elem.value = ret.toString(); // immediate normalization
        }
        return ret;
      }));
  updateLearningRate = (val: number) => this.reduxSet.setMlPolynomialLearningRate(val);
  updateOptimizer = (val: string) => this.reduxSet.setMlPolynomialOptimizer(val);

  resetTrained = () => this.reduxSet.setMlPolynomialFactorsTrained(this.redux.state.ui.ai.mlPolynomial.factorsTrained.map(ii => 0));
  generatePoints = () => this.triggerDoPoints$.next();

  async train(steps: number) {
    if (this.isBusy$.value) {
      return;
    }
    this.isBusy$.next(true);
    try {
      const state = this.redux.state.ui.ai.mlPolynomial;
      const trained = await detectPolynom({
        initialWeights: state.factorsTrained,
        learningRate: state.learningRate,
        loops: steps,
        optimizer: state.optimizer,
        xyFlatData: state.pointsCurrent
      });
      this.reduxSet.setMlPolynomialFactorsTrained(trained);
    } finally {
      this.isBusy$.next(false);
    }
  }

  private doGeneratePoints = () =>
    this.reduxSet.setMlPolynomialPointsCurrent(
      generatePolynomialPoints({
        weights: this.redux.state.ui.ai.mlPolynomial.factorsCurrent,
        points: this.numPoints,
        xFrom: 0,
        xTo: this.xRange
      }));
}
