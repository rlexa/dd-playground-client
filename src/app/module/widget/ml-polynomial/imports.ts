import { CommonModule } from '@angular/common';
import { MatButtonModule, MatCardModule, MatFormFieldModule, MatInputModule, MatSelectModule } from '@angular/material';
import { RouterModule } from '@angular/router';
import { D3numberPipeModule } from 'app/module/pipe/d3number';
import { DiagramPolynomModule } from 'app/module/widget/diagram-polynom';
import { SimpleViewModule } from 'app/module/widget/simple-view';
import { FlexboxModule } from 'dd-flexbox';

export const imports = [
  CommonModule, RouterModule, MatButtonModule, MatCardModule, MatFormFieldModule, MatInputModule, MatSelectModule, FlexboxModule,
  D3numberPipeModule, DiagramPolynomModule, SimpleViewModule,
];
