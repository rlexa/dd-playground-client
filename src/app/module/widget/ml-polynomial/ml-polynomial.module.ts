import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { ROUTE_ROOT, ROUTE_WILDCARD } from 'app/routing';
import { imports } from './imports';
import { MlPolynomialComponent } from './ml-polynomial.component';

const ROUTING = <Routes>[
  { path: ROUTE_ROOT, component: MlPolynomialComponent },
  { path: ROUTE_WILDCARD, redirectTo: ROUTE_ROOT, pathMatch: 'full' },
];

@NgModule({ imports: [...imports, RouterModule.forChild(ROUTING)], exports: [MlPolynomialComponent], declarations: [MlPolynomialComponent] })
export class MlPolynomialModule { }
