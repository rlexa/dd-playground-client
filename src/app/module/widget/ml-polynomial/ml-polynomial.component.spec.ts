import { async, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { provideRedux } from 'app/redux/test';
import { imports } from './imports';
import { MlPolynomialComponent } from './ml-polynomial.component';

describe('MlPolynomialComponent', () => {
  beforeEach(async(() => {
    TestBed.configureCompiler(<any>{ preserveWhitespaces: false }).configureTestingModule({
      imports: [NoopAnimationsModule, RouterTestingModule, ...imports], declarations: [MlPolynomialComponent], providers: [...provideRedux],
    }).compileComponents();
  }));

  test('is created', () => {
    const fixture = TestBed.createComponent(MlPolynomialComponent);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
