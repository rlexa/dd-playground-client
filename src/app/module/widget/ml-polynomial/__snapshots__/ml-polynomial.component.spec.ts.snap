// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MlPolynomialComponent is created 1`] = `
<app-ml-polynomial
  colors={[Function Array]}
  doGeneratePoints={[Function Function]}
  done$={[Function DoneSubject]}
  factorRange={[Function Number]}
  factorsCurrent$={[Function AnonymousSubject]}
  factorsTrained$={[Function BehaviorSubject]}
  generatePoints={[Function Function]}
  isBusy$={[Function BehaviorSubject]}
  learningRate$={[Function BehaviorSubject]}
  numPoints={[Function Number]}
  optimizer$={[Function BehaviorSubject]}
  optimizerDef={[Function String]}
  optimizers={[Function Array]}
  pointsCurrent$={[Function BehaviorSubject]}
  redux={[Function ReduxService]}
  reduxSet={[Function ReduxSetUiAiService]}
  resetTrained={[Function Function]}
  tfjs={[Function TfjsService]}
  tfjsBackend$={[Function BehaviorSubject]}
  tfjsMemory$={[Function BehaviorSubject]}
  trackByIndex={[Function Function]}
  triggerDoPoints$={[Function Subject]}
  updateFactor={[Function Function]}
  updateLearningRate={[Function Function]}
  updateOptimizer={[Function Function]}
  xRange={[Function Number]}
>
  <div
    ddflexbox="vys"
    ng-reflect-dd-flexbox="vys"
    style="display: flex; flex-direction: column; justify-content: flex-start; align-items: flex-start; align-content: stretch; flex-wrap: no-wrap;"
  >
    <mat-card
      class="mat-card"
      ddflexitem="at"
      ng-reflect-dd-flexitem="at"
    >
      <mat-card-title
        class="mat-card-title"
      >
        Polynomial
      </mat-card-title>
      <mat-card-content
        class="mat-card-content"
      >
        <div
          ddflexbox="vys"
          ng-reflect-dd-flexbox="vys"
          style="display: flex; flex-direction: column; justify-content: flex-start; align-items: flex-start; align-content: stretch; flex-wrap: no-wrap;"
        >
          <div
            class="childv"
            ddflexbox="hys"
            ng-reflect-dd-flexbox="hys"
            style="display: flex; flex-direction: row; justify-content: flex-start; align-items: flex-start; align-content: stretch; flex-wrap: no-wrap;"
          >
            
            <table
              class="ng-star-inserted"
            >
              <tbody>
                <tr>
                  <td />
                  
                  <td
                    class="ng-star-inserted"
                  >
                    <mat-form-field
                      class="width-short-input font-table-input mat-input-save-subspace childh mat-form-field ng-tns-c1-4 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
                    >
                      <div
                        class="mat-form-field-wrapper"
                      >
                        <div
                          class="mat-form-field-flex"
                        >
                          
                          
                          <div
                            class="mat-form-field-infix"
                          >
                            <input
                              aria-invalid="false"
                              aria-required="true"
                              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
                              id="mat-input-1"
                              matinput=""
                              max="10"
                              min="-10"
                              ng-reflect-disabled="false"
                              ng-reflect-required=""
                              ng-reflect-type="number"
                              ng-reflect-value="1"
                              required=""
                              type="number"
                            />
                            <span
                              class="mat-form-field-label-wrapper"
                            >
                              
                            </span>
                          </div>
                          
                        </div>
                        
                        <div
                          class="mat-form-field-underline ng-tns-c1-4 ng-star-inserted"
                        >
                          <span
                            class="mat-form-field-ripple"
                          />
                        </div>
                        <div
                          class="mat-form-field-subscript-wrapper"
                          ng-reflect-ng-switch="hint"
                        >
                          
                          
                          <div
                            class="mat-form-field-hint-wrapper ng-tns-c1-4 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                          >
                            
                            <div
                              class="mat-form-field-hint-spacer"
                            />
                          </div>
                        </div>
                      </div>
                    </mat-form-field>
                  </td>
                  <td
                    class="ng-star-inserted"
                  >
                    <mat-form-field
                      class="width-short-input font-table-input mat-input-save-subspace childh mat-form-field ng-tns-c1-5 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
                    >
                      <div
                        class="mat-form-field-wrapper"
                      >
                        <div
                          class="mat-form-field-flex"
                        >
                          
                          
                          <div
                            class="mat-form-field-infix"
                          >
                            <input
                              aria-invalid="false"
                              aria-required="true"
                              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
                              id="mat-input-2"
                              matinput=""
                              max="10"
                              min="-10"
                              ng-reflect-disabled="false"
                              ng-reflect-required=""
                              ng-reflect-type="number"
                              ng-reflect-value="1"
                              required=""
                              type="number"
                            />
                            <span
                              class="mat-form-field-label-wrapper"
                            >
                              
                            </span>
                          </div>
                          
                        </div>
                        
                        <div
                          class="mat-form-field-underline ng-tns-c1-5 ng-star-inserted"
                        >
                          <span
                            class="mat-form-field-ripple"
                          />
                        </div>
                        <div
                          class="mat-form-field-subscript-wrapper"
                          ng-reflect-ng-switch="hint"
                        >
                          
                          
                          <div
                            class="mat-form-field-hint-wrapper ng-tns-c1-5 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                          >
                            
                            <div
                              class="mat-form-field-hint-spacer"
                            />
                          </div>
                        </div>
                      </div>
                    </mat-form-field>
                  </td>
                  <td
                    class="ng-star-inserted"
                  >
                    <mat-form-field
                      class="width-short-input font-table-input mat-input-save-subspace childh mat-form-field ng-tns-c1-6 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
                    >
                      <div
                        class="mat-form-field-wrapper"
                      >
                        <div
                          class="mat-form-field-flex"
                        >
                          
                          
                          <div
                            class="mat-form-field-infix"
                          >
                            <input
                              aria-invalid="false"
                              aria-required="true"
                              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
                              id="mat-input-3"
                              matinput=""
                              max="10"
                              min="-10"
                              ng-reflect-disabled="false"
                              ng-reflect-required=""
                              ng-reflect-type="number"
                              ng-reflect-value="1"
                              required=""
                              type="number"
                            />
                            <span
                              class="mat-form-field-label-wrapper"
                            >
                              
                            </span>
                          </div>
                          
                        </div>
                        
                        <div
                          class="mat-form-field-underline ng-tns-c1-6 ng-star-inserted"
                        >
                          <span
                            class="mat-form-field-ripple"
                          />
                        </div>
                        <div
                          class="mat-form-field-subscript-wrapper"
                          ng-reflect-ng-switch="hint"
                        >
                          
                          
                          <div
                            class="mat-form-field-hint-wrapper ng-tns-c1-6 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                          >
                            
                            <div
                              class="mat-form-field-hint-spacer"
                            />
                          </div>
                        </div>
                      </div>
                    </mat-form-field>
                  </td>
                  <td
                    class="ng-star-inserted"
                  >
                    <mat-form-field
                      class="width-short-input font-table-input mat-input-save-subspace childh mat-form-field ng-tns-c1-7 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
                    >
                      <div
                        class="mat-form-field-wrapper"
                      >
                        <div
                          class="mat-form-field-flex"
                        >
                          
                          
                          <div
                            class="mat-form-field-infix"
                          >
                            <input
                              aria-invalid="false"
                              aria-required="true"
                              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
                              id="mat-input-4"
                              matinput=""
                              max="10"
                              min="-10"
                              ng-reflect-disabled="false"
                              ng-reflect-required=""
                              ng-reflect-type="number"
                              ng-reflect-value="1"
                              required=""
                              type="number"
                            />
                            <span
                              class="mat-form-field-label-wrapper"
                            >
                              
                            </span>
                          </div>
                          
                        </div>
                        
                        <div
                          class="mat-form-field-underline ng-tns-c1-7 ng-star-inserted"
                        >
                          <span
                            class="mat-form-field-ripple"
                          />
                        </div>
                        <div
                          class="mat-form-field-subscript-wrapper"
                          ng-reflect-ng-switch="hint"
                        >
                          
                          
                          <div
                            class="mat-form-field-hint-wrapper ng-tns-c1-7 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                          >
                            
                            <div
                              class="mat-form-field-hint-spacer"
                            />
                          </div>
                        </div>
                      </div>
                    </mat-form-field>
                  </td>
                </tr>
                
                <tr
                  class="ng-star-inserted"
                  ng-reflect-ng-style="[object Object]"
                  style="color: rgb(0, 0, 255); vertical-align: bottom;"
                >
                  <td
                    class="text-nowrap"
                  >
                    f(x) = 
                  </td>
                  
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +1.0 
                    
                    
                     x 
                    
                    <sup
                      class="font-small ng-star-inserted"
                    >
                      3
                    </sup>
                    
                  </td>
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +1.0 
                    
                    
                     x 
                    
                    <sup
                      class="font-small ng-star-inserted"
                    >
                      2
                    </sup>
                    
                  </td>
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +1.0 
                    
                    
                     x 
                    
                    
                  </td>
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +1.0 
                    
                  </td>
                </tr>
                <tr
                  class="ng-star-inserted"
                  ng-reflect-ng-style="[object Object]"
                  style="color: rgb(255, 100, 100); vertical-align: bottom;"
                >
                  <td
                    class="text-nowrap"
                  >
                    f(x) = 
                  </td>
                  
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +0.0 
                    
                    
                     x 
                    
                    <sup
                      class="font-small ng-star-inserted"
                    >
                      3
                    </sup>
                    
                  </td>
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +0.0 
                    
                    
                     x 
                    
                    <sup
                      class="font-small ng-star-inserted"
                    >
                      2
                    </sup>
                    
                  </td>
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +0.0 
                    
                    
                     x 
                    
                    
                  </td>
                  
                  <td
                    class="ng-star-inserted"
                  >
                      +0.0 
                    
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="childhv"
              ddflexbox="vys"
              ng-reflect-dd-flexbox="vys"
              style="display: flex; flex-direction: column; justify-content: flex-start; align-items: flex-start; align-content: stretch; flex-wrap: no-wrap;"
            >
              <div
                class="childhv"
                ddflexbox="hyscsw"
                ng-reflect-dd-flexbox="hyscsw"
                style="display: flex; flex-direction: row; justify-content: flex-start; align-items: flex-start; align-content: flex-start; flex-wrap: wrap;"
              >
                <div
                  class="childh"
                >
                  <button
                    class="mat-raised-button _mat-animation-noopable"
                    mat-raised-button=""
                    ng-reflect-disabled="false"
                  >
                    <span
                      class="mat-button-wrapper"
                    >
                      NEW POINTS
                    </span>
                    <div
                      class="mat-button-ripple mat-ripple"
                      matripple=""
                      ng-reflect-centered="false"
                      ng-reflect-disabled="false"
                      ng-reflect-trigger="[object HTMLButtonElement]"
                    />
                    <div
                      class="mat-button-focus-overlay"
                    />
                  </button>
                </div>
                <div
                  class="childh"
                >
                  <button
                    class="mat-raised-button _mat-animation-noopable"
                    mat-raised-button=""
                    ng-reflect-disabled="false"
                  >
                    <span
                      class="mat-button-wrapper"
                    >
                      RESET TRAINED
                    </span>
                    <div
                      class="mat-button-ripple mat-ripple"
                      matripple=""
                      ng-reflect-centered="false"
                      ng-reflect-disabled="false"
                      ng-reflect-trigger="[object HTMLButtonElement]"
                    />
                    <div
                      class="mat-button-focus-overlay"
                    />
                  </button>
                </div>
              </div>
              <div
                class="childhv"
                ddflexbox="hyscsw"
                ng-reflect-dd-flexbox="hyscsw"
                style="display: flex; flex-direction: row; justify-content: flex-start; align-items: flex-start; align-content: flex-start; flex-wrap: wrap;"
              >
                <div
                  class="margin-top"
                  ddflexbox="hys"
                  ng-reflect-dd-flexbox="hys"
                  style="display: flex; flex-direction: row; justify-content: flex-start; align-items: flex-start; align-content: stretch; flex-wrap: no-wrap;"
                >
                  <label
                    class="childhv"
                  >
                    TFJS:
                  </label>
                  <app-simple-view
                    class="ng-tns-c3-0"
                    ng-reflect-data="[object Object]"
                    ng-reflect-is-dense="true"
                    ng-reflect-is-expandable="true"
                    ng-reflect-is-expanded="false"
                  >
                    
                    <div
                      class="position-relative ng-tns-c3-0 ng-star-inserted"
                    >
                      <div
                        class="scroll-x-auto"
                        ddflexbox="hys"
                        ng-reflect-dd-flexbox="hys"
                        style="max-width: 100%; display: flex; flex-direction: row; justify-content: flex-start; align-items: flex-start; align-content: stretch; flex-wrap: no-wrap;"
                      >
                        
                        <button
                          class="inline-button mat-button ng-tns-c3-0 _mat-animation-noopable ng-star-inserted"
                          mat-button=""
                        >
                          <span
                            class="mat-button-wrapper"
                          >
                            <i
                              class="material-icons rot-90 ng-tns-c3-0"
                            >
                              expand_more
                            </i>
                          </span>
                          <div
                            class="mat-button-ripple mat-ripple"
                            matripple=""
                            ng-reflect-centered="false"
                            ng-reflect-disabled="false"
                            ng-reflect-trigger="[object HTMLButtonElement]"
                          />
                          <div
                            class="mat-button-focus-overlay"
                          />
                        </button>
                        
                      </div>
                    </div>
                  </app-simple-view>
                </div>
                <mat-form-field
                  class="childh width-short-input font-table-input mat-form-field ng-tns-c1-1 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <input
                          aria-invalid="false"
                          aria-required="false"
                          class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
                          id="mat-input-0"
                          matinput=""
                          min="0.01"
                          ng-reflect-disabled="false"
                          ng-reflect-placeholder="Rate"
                          ng-reflect-type="number"
                          ng-reflect-value="0.1"
                          placeholder="Rate"
                          step=".1"
                          type="number"
                        />
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                          <label
                            aria-owns="mat-input-0"
                            class="mat-form-field-label ng-tns-c1-1 ng-star-inserted"
                            for="mat-input-0"
                            id="mat-form-field-label-1"
                            ng-reflect-disabled="true"
                            ng-reflect-ng-switch="false"
                          >
                            
                            
                            Rate
                            
                            
                          </label>
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-1 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
                <mat-form-field
                  class="childh width-shorter-input font-table-input mat-form-field ng-tns-c1-2 mat-primary mat-form-field-type-mat-select mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <mat-select
                          aria-disabled="false"
                          aria-invalid="false"
                          aria-label="Optimizer"
                          aria-multiselectable="false"
                          aria-required="false"
                          class="mat-select ng-tns-c4-3 mat-select-empty ng-star-inserted"
                          id="mat-select-0"
                          ng-reflect-placeholder="Optimizer"
                          ng-reflect-value="adam"
                          placeholder="Optimizer"
                          role="listbox"
                          tabindex="0"
                        >
                          <div
                            aria-hidden="true"
                            cdk-overlay-origin=""
                            class="mat-select-trigger"
                          >
                            <div
                              class="mat-select-value"
                              ng-reflect-ng-switch="true"
                            >
                              
                              <span
                                class="mat-select-placeholder ng-tns-c4-3 ng-star-inserted"
                              >
                                Optimizer
                              </span>
                              
                            </div>
                            <div
                              class="mat-select-arrow-wrapper"
                            >
                              <div
                                class="mat-select-arrow"
                              />
                            </div>
                          </div>
                          
                        </mat-select>
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                          <label
                            aria-owns="mat-select-0"
                            class="mat-form-field-label ng-tns-c1-2 mat-empty mat-form-field-empty ng-star-inserted"
                            for="mat-select-0"
                            id="mat-form-field-label-3"
                            ng-reflect-disabled="true"
                            ng-reflect-ng-switch="false"
                          >
                            
                            
                            Optimizer
                            
                            
                          </label>
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-2 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-2 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
                
                <button
                  class="auto-min-width mat-raised-button _mat-animation-noopable ng-star-inserted"
                  mat-raised-button=""
                  ng-reflect-disabled="false"
                >
                  <span
                    class="mat-button-wrapper"
                  >
                     TRAIN x1 
                  </span>
                  <div
                    class="mat-button-ripple mat-ripple"
                    matripple=""
                    ng-reflect-centered="false"
                    ng-reflect-disabled="false"
                    ng-reflect-trigger="[object HTMLButtonElement]"
                  />
                  <div
                    class="mat-button-focus-overlay"
                  />
                </button>
                <button
                  class="auto-min-width mat-raised-button _mat-animation-noopable ng-star-inserted"
                  mat-raised-button=""
                  ng-reflect-disabled="false"
                >
                  <span
                    class="mat-button-wrapper"
                  >
                     x5 
                  </span>
                  <div
                    class="mat-button-ripple mat-ripple"
                    matripple=""
                    ng-reflect-centered="false"
                    ng-reflect-disabled="false"
                    ng-reflect-trigger="[object HTMLButtonElement]"
                  />
                  <div
                    class="mat-button-focus-overlay"
                  />
                </button>
                <button
                  class="auto-min-width mat-raised-button _mat-animation-noopable ng-star-inserted"
                  mat-raised-button=""
                  ng-reflect-disabled="false"
                >
                  <span
                    class="mat-button-wrapper"
                  >
                     x10 
                  </span>
                  <div
                    class="mat-button-ripple mat-ripple"
                    matripple=""
                    ng-reflect-centered="false"
                    ng-reflect-disabled="false"
                    ng-reflect-trigger="[object HTMLButtonElement]"
                  />
                  <div
                    class="mat-button-focus-overlay"
                  />
                </button>
                <button
                  class="auto-min-width mat-raised-button _mat-animation-noopable ng-star-inserted"
                  mat-raised-button=""
                  ng-reflect-disabled="false"
                >
                  <span
                    class="mat-button-wrapper"
                  >
                     x100 
                  </span>
                  <div
                    class="mat-button-ripple mat-ripple"
                    matripple=""
                    ng-reflect-centered="false"
                    ng-reflect-disabled="false"
                    ng-reflect-trigger="[object HTMLButtonElement]"
                  />
                  <div
                    class="mat-button-focus-overlay"
                  />
                </button>
                <button
                  class="auto-min-width mat-raised-button _mat-animation-noopable ng-star-inserted"
                  mat-raised-button=""
                  ng-reflect-disabled="false"
                >
                  <span
                    class="mat-button-wrapper"
                  >
                     x1000 
                  </span>
                  <div
                    class="mat-button-ripple mat-ripple"
                    matripple=""
                    ng-reflect-centered="false"
                    ng-reflect-disabled="false"
                    ng-reflect-trigger="[object HTMLButtonElement]"
                  />
                  <div
                    class="mat-button-focus-overlay"
                  />
                </button>
              </div>
            </div>
          </div>
          <app-diagram-polynom
            ddflexitem="at"
            ng-reflect-dd-flexitem="at"
            ng-reflect-points=""
            ng-reflect-polynomial-colors="rgba(0, 0, 255, 200),rgba(255,"
            ng-reflect-polynomials="1,1,1,1,0,0,0,0"
            style="min-width: 60rem;"
          >
            <div
              class="svg-container position-relative"
            >
              <svg
                id="svgChart"
                preserveAspectRatio="xMinYMin meet"
                viewBox="0 0 55 375"
              >
                <g
                  transform="translate(35, 5)"
                >
                  <rect
                    class="filler"
                    height="100%"
                    style="fill: transparent;"
                    width="100%"
                  />
                  <g
                    class="content"
                  />
                  <g
                    class="axis-x"
                    transform="translate(0,350)"
                  />
                  <g
                    class="axis-y"
                  />
                </g>
              </svg>
            </div>
          </app-diagram-polynom>
        </div>
      </mat-card-content>
    </mat-card>
  </div>
</app-ml-polynomial>
`;
